## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [26]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [27]:
# TODO: load `data/processed/shopping_cleaned.csv` as a pandas dataframe

df = pd.read_csv("../data/processed/shopping_cleaned.csv")

In [28]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

df_pivot=pd.pivot_table(df, values='Location', index='Color', columns='Season', aggfunc=len, fill_value=0)

In [29]:
# TODO: Display this pivot table
print(df_pivot)


Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine               83       2      13     111
Baby blue               16      69      84      21
Black                   84       5       9     124
Brick red               88       7      12     110
Brown                   97       5       3      89
Burnt orange            93       4       5     112
Fuchsia                  3      45      78      18
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Mauve                   86       5       7      95
Muted mustard yellow    84       8      13      87
Pale peach               9      52      81      15
Periwinkle               9      54      82       8
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Turquoise                7      49      74      16
White                   10      63      81      11


In [30]:
# TODO: Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

fall_column = df_pivot['Fall'].sort_values(ascending=False)

print(fall_column)

Color
Brown                   97
Burnt orange            93
Terra cotta             91
Ruby Red                90
Brick red               88
Mauve                   86
Black                   84
Muted mustard yellow    84
Aubergine               83
Baby blue               16
Lavender                13
White                   10
Periwinkle               9
Pale peach               9
Lemon yellow             9
Turquoise                7
Fuchsia                  3
Name: Fall, dtype: int64


In [31]:
# TODO: Select the "Winter" column from this pivot table and display the sort order in descending order

winter_column = df_pivot['Winter'].sort_values(ascending=False)

print(winter_column)

Color
Black                   124
Burnt orange            112
Aubergine               111
Brick red               110
Ruby Red                107
Terra cotta             101
Mauve                    95
Brown                    89
Muted mustard yellow     87
Baby blue                21
Fuchsia                  18
Turquoise                16
Pale peach               15
Lemon yellow             13
White                    11
Periwinkle                8
Lavender                  6
Name: Winter, dtype: int64


In [32]:
# TODO: Select the "Spring" column from this pivot table and display the sort order in descending order

spring_column = df_pivot['Spring'].sort_values(ascending=False)

print(spring_column)

Color
Baby blue               69
White                   63
Lavender                55
Periwinkle              54
Pale peach              52
Turquoise               49
Lemon yellow            47
Fuchsia                 45
Muted mustard yellow     8
Brick red                7
Mauve                    5
Brown                    5
Black                    5
Burnt orange             4
Terra cotta              4
Ruby Red                 3
Aubergine                2
Name: Spring, dtype: int64


In [33]:
# TODO: Select the "Summer" column from this pivot table and display the sort order in descending order

summer_column = df_pivot['Summer'].sort_values(ascending=False)

print(summer_column)

Color
Lavender                102
Lemon yellow             98
Baby blue                84
Periwinkle               82
Pale peach               81
White                    81
Fuchsia                  78
Turquoise                74
Terra cotta              15
Muted mustard yellow     13
Aubergine                13
Brick red                12
Black                     9
Ruby Red                  8
Mauve                     7
Burnt orange              5
Brown                     3
Name: Summer, dtype: int64


In [34]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html
df_pivot_idmax=df_pivot.idxmax()
print(df_pivot_idmax)


Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object


In [35]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

df_pivot2=pd.pivot_table(df, values='Location', index='Item Purchased', columns='Season', aggfunc=len, fill_value=0)
print(df_pivot2)

Season          Fall  Spring  Summer  Winter
Item Purchased                              
Backpack          72       0       0     113
Belt              25       0       0       0
Boots             51       0       0      83
Dress             36      74     118       0
Gloves            68       0       0     112
Handbag           31      78     111       0
Hoodie            58       0       0     105
Jacket            47       0       0      87
Jeans             46       0       0       0
Leggings          58       0       0     118
Pajamas           65       0       0     111
Running Shoes     25      89     134       0
Scarf             56       0       0     101
Shorts            31      76     152       0
Socks             71       0       0     101
Sunglasses        43      80     135       0
Sweater           72       0       0     113
T-shirt           17      80     115       0


In [36]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html
df_pivot2_idmax=df_pivot2.idxmax()
print(df_pivot2_idmax)


Season
Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object


In [37]:
# TODO: Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two dataframes into seperate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

promo_code_users = df[df['Promo Code Used'] == 'Yes']

non_promo_code_users = df[df['Promo Code Used'] == 'No']

print("DataFrame for Promo Code Users:")
print(promo_code_users.head())

print("DataFrame for Non-Promo Code Users:")
print(non_promo_code_users.head())

DataFrame for Promo Code Users:
       Age  Gender Item Purchased  Purchase Amount (USD)       Location Size  \
1785  36.0  Female  Running Shoes              62.442101  Massachusetts    L   
1786  57.0    Male       Backpack              51.969839       Michigan    M   
1787  26.0    Male          Socks              54.899050      Wisconsin    M   
1788  36.0  Female        Handbag              46.883976       Delaware    S   
1789  34.0  Female       Backpack              45.171082          Texas    L   

          Color  Season Review Rating   Shipping Type Promo Code Used  \
1785      White  Summer       Present        Standard             Yes   
1786  Aubergine    Fall       Present  2-Day Shipping             Yes   
1787      Black    Fall       Present  2-Day Shipping             Yes   
1788      Mauve  Summer       Present  2-Day Shipping             Yes   
1789      Mauve  Winter       Present  2-Day Shipping             Yes   

      Previous Purchases Payment Method  
1785  

In [38]:
# TODO: Print the shape of the promo code users dataframe. This should be (1373, 13)

promo_code_users.shape

(1373, 13)

In [39]:
# TODO: Print the average purchase size of promo code users dataframe. This should be 50.06747998236351

average_purchase_size = promo_code_users['Purchase Amount (USD)'].mean()
print(average_purchase_size)

50.06747998236351


In [40]:
# TODO: Print the shape of the non promo code users dataframe. This should be (1785, 13)

non_promo_code_users.shape


(1785, 13)

In [41]:
# TODO: Print the average purchase size of non promo code users dataframe. This should be 30.155750765104113

average_purchase_size = non_promo_code_users['Purchase Amount (USD)'].mean()
print(average_purchase_size)

30.155750765104113


Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

promo-code will affect the amount spending between promo-code users and non-promo-code users.  

Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

promo-code will not affect the amount spending between promo-code users and non-promo-code users. 

In [42]:
# TODO: Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

from scipy.stats import ttest_ind


In [43]:
t_statistic = ttest_ind(promo_code_users['Purchase Amount (USD)'], non_promo_code_users['Purchase Amount (USD)'])
t_statistic

TtestResult(statistic=93.2464714350831, pvalue=0.0, df=3156.0)

In [44]:
# TODO: Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

g_review=df.groupby('Review Rating')

In [45]:
# TODO: Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html

g_review.groups.keys()


dict_keys(['Present'])

In [46]:
# TODO: calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html
columns_to_average = ["Purchase Amount (USD)", "Age", "Previous Purchases"]
average_values = g_review[columns_to_average].mean()
print("Average values for each group:")
print(average_values)


Average values for each group:
               Purchase Amount (USD)        Age  Previous Purchases
Review Rating                                                      
Present                     38.81275  34.185244            5.860988


In [47]:
# TODO: Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

df['Prev Purchase Group'] = pd.cut(df['Previous Purchases'], bins=2, labels=bin_labels, include_lowest=True)

print(df.head())



NameError: name 'bin_labels' is not defined

In [ ]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

df[df["Prev Purchase Group"]=="Low"]["Review Rating"].value_counts()

Review Rating
Missing    1686
Present     840
Name: count, dtype: int64

In [ ]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

df[df["Prev Purchase Group"]=="High"]["Review Rating"].value_counts()

Review Rating
Missing    334
Present    298
Name: count, dtype: int64

## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

Fall top three colors: 
Brown                   97
Burnt orange            93
Terra cotta             91
Winter top three colors
Black                   124
Burnt orange            112
Aubergine               111

### Q2

What are the top three colors for Spring & Summer?

Spring:
Color
Baby blue               69
White                   63
Lavender                55
Summer: 
Color
Lavender                102
Lemon yellow             98
Baby blue                84


### Q3

What is the most popular clothing item by season?

Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

t-statistic of 93.2464714350831. If the p-value is sufficiently low (e.g., less than 0.05), you would likely reject the null hypothesis.


93.2464714350831

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

high group. High group has more loyal customers. They tend to shop more. 